### PR TITLE
feat: implemented days of supply edc integration

### PR DESCRIPTION
**Pull request by ReneSchroederLJ**
## Description

- add SAMM models for Days of Supply
- implemented logic for registering the corresponding Assets and Contract
- implemented the request service for exchanging Days of Supply through the EDC

reolves #312 

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [x] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files
- [x] If helm chart has been changed, the chart version has been bumped to either next major, minor or patch level (compared to released chart).

### CODE REVIEW

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/service/DaysOfSupplyRequestApiService.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            return null;
+        }
+        MaterialPartnerRelation mpr = switch (direction) {
+            case OUTBOUND -> mprService.findAll().stream()
+                    .filter(m -> m.getPartner().equals(partner) && m.getMaterial().getMaterialNumberCx().equals(materialNumberCx))
+                    .findFirst().orElse(null);
+            case INBOUND -> mprService.findByPartnerAndPartnerCXNumber(partner, materialNumberCx);
+        };
+        if (mpr == null) {
+            log.error("Could not identify Material-Partner Relation with material number " + materialNumberCx + " and partner bpnl " + partner.getBpnl());
```
[the remark this answers is not in the document]
Comment (ReneSchroederLJ) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/service/DaysOfSupplyRequestApiService.java:80]: The value for `materialNumberCx` is verified as URN or UUID before use and should not pose a risk.

**Review by tom-rm-meyer-ISST — CHANGES_REQUESTED** [review]
Got some smaller changes. Likely we should instead of solving a few things open issues to do a refactoring.

The updates broke the following test cases of the bruno collection
- Test_02-EDC/Check from Customer POV/Query catalog for Submodel Assets
- Test_02-EDC/Check from Supplier POV/Query catalog for Submodel Assets 
The test, Test_01-MAD/01_01-Init/Customer/MAD/Get All Partners has been missed by me during a recent fix.
- Also the ShellDescriptors should be checked regarding the new submodels

I'm wondering if this should also be handled with a separate pr and if we should do the PR also against a feature branch. Then we could handle the quick things in this pr. Additional stuff (update bruno collection, some javadoc) on the branch, too and we just merge the feature fully. Feel free to also reach out to me via MS teams to quickly align here.

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/domain/model/AssetType.java [hunk 2]
```diff
@@ -27,6 +27,7 @@ public enum AssetType {
     DEMAND_SUBMODEL("urn:samm:io.catenax.short_term_material_demand:1.0.0#ShortTermMaterialDemand", "$value", "ShortTermMaterialDemand", "1.0"),
     DELIVERY_SUBMODEL("urn:samm:io.catenax.delivery_information:2.0.0#DeliveryInformation", "$value", "DeliveryInformation", "2.0"),
     NOTIFICATION("urn:samm:io.catenax.demand_and_capacity_notification:2.0.0#DemandAndCapacityNotification", "none", "none", "2.0"),
+    DAYS_OF_SUPPLY("urn:samm:io.catenax.days_of_supply:2.0.0#DaysOfSupply", "$value", "DaysOfSupply", "1.0"),
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/domain/model/AssetType.java:30]:
```suggestion
    DAYS_OF_SUPPLY("urn:samm:io.catenax.days_of_supply:2.0.0#DaysOfSupply", "$value", "DaysOfSupply", "2.0"),
```

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/service/CustomerSupplyService.java [hunk 3]
```diff
@@ -108,18 +104,13 @@ public final List<ReportedCustomerSupply> findAllByFilters(Optional<String> ownM
...
-
     public boolean validate(ReportedCustomerSupply daysOfSupply) {
         return 
             daysOfSupply.getPartner() != null &&
             daysOfSupply.getMaterial() != null &&
             daysOfSupply.getDate() != null &&
             daysOfSupply.getStockLocationBPNS() != null &&
             daysOfSupply.getPartner() != partnerService.getOwnPartnerEntity() &&
-            daysOfSupply.getPartner().getSites().stream().anyMatch(site -> site.getBpns().equals(daysOfSupply.getStockLocationBPNS())) &&
-            (daysOfSupply.getStockLocationBPNA().equals(null) || daysOfSupply.getStockLocationBPNA().equals(daysOfSupply.getStockLocationBPNS()));
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/service/CustomerSupplyService.java:123]: you;re right. Good catch (Y)

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/adapter/DaysOfSupplySammMapper.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        if (mpr == null) {
+            log.warn("Could not identify materialPartnerRelation with ownMaterialNumber " + material.getOwnMaterialNumber()
+                    + " and partner bpnl " + partner.getBpnl());
+            return null;
+        }
+        samm.setMaterialGlobalAssetId(material.getMaterialNumberCx());
+        samm.setDirection(DirectionCharacteristic.OUTBOUND);
+        var suppliesPerDay = new HashSet<AllocatedDaysOfSupply>();
+        samm.setAllocatedDaysOfSupply(suppliesPerDay);
+        for (var supplyBySite : suppliesBySite) {
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/adapter/DaysOfSupplySammMapper.java:70]: should have removed that.

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/adapter/DaysOfSupplySammMapper.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                supplyBySite.stream().map(sup -> new QuantityOfDaysOfSupply(sup.getDate(), sup.getDaysOfSupply())).toList());
+            suppliesPerDay.add(allocatedDaysOfSupply);
+        }
+        return samm;
+    }
+
+    public List<ReportedCustomerSupply> sammToReportedCustomerSupply(DaysOfSupply samm, Partner partner) {
+        String matNbrCatenaX = samm.getMaterialGlobalAssetId();
+        ArrayList<ReportedCustomerSupply> outputList = new ArrayList<>();
+        var material = materialService.findByMaterialNumberCx(matNbrCatenaX);
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/supply/logic/adapter/DaysOfSupplySammMapper.java:117]: should have removed that.

**Review by tom-rm-meyer-ISST — CHANGES_REQUESTED** [review]
Thanks for incorporating. Just a small version change missing. You can directly commit via suggestion! Re-request review for merge

**Review by tom-rm-meyer-ISST — APPROVED** [review]
LGTM, issues for remaining comments will be creaed by @ReneSchroederLJ (feel free to link the pr and close the issues).

Helm-Chart Link is ignored on purpose. Version Bump will be done during release.

**Pull request merged**